src/adr/0018-polyglot-strategy.md

```diff
@@ -0,0 +1,158 @@
+# 16. Polyglot codebase
+
+Date: 2020-02-03
+
+## Status
+
+Proposed
+
+## Context & Problem Statement
+
+James is written in Java for a very long time. In recent years, Java modernized a lot after a decade of slow progress.
+
+However, in the meantime, most software relying on the JVM started supporting alternative JVM languages to keep being relevant.
+
+It includes Groovy, Clojure, Scala and more recently Kotlin.
+
+Not being open to those alternative languages can be a problem for James adoption.
+
+## Decision drivers
+
+Nowadays, libraries and framework targeting the JVM are expected to support usage of one or several of these alternative languages.
+
+James being not only a mail server but also a development framework needs to reach those expectations.
+
+At the same time, more and more developers and languages adopt Function Programming (FP) idioms to solve their problems.
+
+## Considered options
+
+1. Let the users figure out how to make polyglot setups
+2. Document the usage of polyglot mailets for some popular languages
+3. Document the usage of polyglot components for some popular languages
+4. Actually implement some mailets in some popular languages
+5. Actually implement some components in some popular languages
+
+About Languages:
+
+I. Clojure
+II. Groovy
+III. Kotlin
+IV. Scala
+
+## Decision
+
+We decided on options 4, 5 and IV.
```

Comment:
IMO we should act 2.

Reply:
IMO having some existing mailets is the best documentation for developers (see https://github.com/linagora/james-project/pull/1742 that we didn't merged but should revive)
We can of course do better but as a developer, if I can just copy/paste some existing code and start straight away, I'm happy.